test/test_utils/ad_utils.jl

```diff
@@ -0,0 +1,213 @@
+module ADUtils
+
+import ForwardDiff
+import ReverseDiff
+import Test
+import Tracker
+import Turing
+import Turing: DynamicPPL
+import Zygote
+
+export ADTypeCheckContext
+
+"""Element types that are always valid for a VarInfo regardless of ADType."""
+const always_valid_eltypes = (AbstractFloat, AbstractIrrational, Integer, Rational)
+
+"""A dictionary mapping ADTypes to the element types they use."""
+const eltypes_by_adtype = Dict(
+    Turing.AutoForwardDiff => (ForwardDiff.Dual,),
+    Turing.AutoReverseDiff => (
+        ReverseDiff.TrackedArray,
+        ReverseDiff.TrackedMatrix,
+        ReverseDiff.TrackedReal,
+        ReverseDiff.TrackedStyle,
+        ReverseDiff.TrackedType,
+        ReverseDiff.TrackedVecOrMat,
+        ReverseDiff.TrackedVector,
+    ),
+    # TODO(mhauru) Zygote.Dual is actually the same as ForwardDiff.Dual, so can't
+    # distinguish between the two.
+    Turing.AutoZygote => (Zygote.Dual,),
```

Comment:
Yeah there's no way you can do this here unfortunately 😕 

Possibly crazy idea, but it might make sense to specifically overload the _adjoint_ computation for a given backend which doesn't use types, e.g.
```julia
Zygote.@adjoint function check_adtype(...)
    # Should only be hit if we're using Zygote.jl.
    ...
end
```
? Could do the same with Enzyme.jl and Tapir.jl, both of which would suffer from the same issue

Reply:
I like this, and implemented it. Turns out there's `Zygote.hook` for exactly these sorts of things.

I thought a bit about switching to doing all checks like this, and giving up on the element type approach, but that seemed like it would get more complicated than is worth at this point. We now catch all cases except if one uses ForwardDiff when Zygote is expected, which seems sufficient to me.

Reply:
Very, very nice @mhauru :)